Mathlib/Tactic/Linter/Multigoal.lean

```diff
@@ -0,0 +1,161 @@
+    ``Lean.Parser.Tactic.skip,
+    `Batteries.Tactic.«tacticOn_goal-_=>_»,
+    `Mathlib.Tactic.«tacticSwap_var__,,»,
+    -- tactic combinators
+    ``Lean.Parser.Tactic.tacticRepeat_,
+    ``Lean.Parser.Tactic.tacticTry_,
+    -- creating new goals
+    ``Lean.Parser.Tactic.paren,
+    ``Lean.Parser.Tactic.case,
+    ``Lean.Parser.Tactic.constructor,
+    `Mathlib.Tactic.tacticAssumption',
+    ``Lean.Parser.Tactic.induction,
+    ``Lean.Parser.Tactic.cases,
+    ``Lean.Parser.Tactic.intros,
+    ``Lean.Parser.Tactic.injections,
+    ``Lean.Parser.Tactic.substVars,
+    `Batteries.Tactic.«tacticPick_goal-_»,
+    ``Lean.Parser.Tactic.case',
+    `«tactic#adaptation_note_»,
+    `tacticSleep_heartbeats_
+  ]
+
+/-- The `SyntaxNodeKind`s in `ignoreBranch` correspond to tactics that disable the linter from
+their first application until the corresponding proof branch is closed.
+Reasons for ignoring these tactics include
+* the linter gets confused by the proof management, e.g. `conv`;
+* the tactics are *intended* to act on multiple goals, e.g. `repeat`, `any_goals`, `all_goals`, ...
+
+There is some overlap in scope between `exclusions` and `ignoreBranch`.
+-/
+abbrev ignoreBranch : Std.HashSet SyntaxNodeKind := .ofList [
+    ``Lean.Parser.Tactic.Conv.conv,
+    `Mathlib.Tactic.Conv.convLHS,
+    `Mathlib.Tactic.Conv.convRHS,
+    ``Lean.Parser.Tactic.first,
+    ``Lean.Parser.Tactic.repeat',
+    ``Lean.Parser.Tactic.tacticIterate____,
+    ``Lean.Parser.Tactic.anyGoals,
+    ``Lean.Parser.Tactic.allGoals,
+    ``Lean.Parser.Tactic.focus
+  ]
+
+/-- `getManyGoals t` returns the syntax nodes of the `InfoTree` `t` corresponding to tactic calls
+which
+* leave at least one goal that was not present before it ran;
+* are not excluded through `exclusions` or `ignoreBranch`.
+-/
+partial
+def getManyGoals : InfoTree → Array (Syntax × Nat)
+  | .node info args =>
+    let kargs := (args.map getManyGoals).toArray.flatten
+    if let .ofTacticInfo info := info then
+      if ignoreBranch.contains info.stx.getKind then #[] else
+      if let .original .. := info.stx.getHeadInfo then
+        let newGoals := info.goalsAfter.filter (info.goalsBefore.contains ·)
+        if newGoals.length != 0 && !exclusions.contains info.stx.getKind then
+          kargs.push (info.stx, newGoals.length)
+        else kargs
+      else kargs
+    else kargs
+  | .context _ t => getManyGoals t
+  | _ => default
+
+@[inherit_doc Mathlib.Linter.linter.style.multiGoal]
+def multiGoalLinter : Linter where
+  run := withSetOptionIn fun _stx => do
+    unless Linter.getLinterValue linter.style.multiGoal (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    let trees ← getInfoTrees
+    for t in trees.toArray do
+      for (s, n) in getManyGoals t do
+        Linter.logLint linter.style.multiGoal s
+          m!"There are {n+1} unclosed goals before '{s}' and at least one remaining goal after.\n\
```

Comment:
Not a native speaker, but: is "afterwards" better English?

Reply:
I don't know, both sound correct to me, so I went with your version!